src/h3/mod.rs

```diff
@@ -26,4 +26,778 @@
+    WrongStreamDirection = -18,
+
+    /// Inform client that remainder of request is not needed. Used in
+    /// STOP_SENDING only.
+    EarlyResponse        = -19,
+
+    /// No SETTINGS frame at beggining of control stream.
+    MissingSettings      = -20,
+
+    /// A frame was received which is not permitted in the current state.
+    UnexpectedFrame      = -21,
+
+    /// Server rejected request without performing any application processing.
+    RequestRejected      = -22,
+
+    /// Peer violated protocol requirements in a way that doesn't match a more
+    /// specific code.
+    GeneralProtocolError = -23,
+
+    /// TODO: malformed frame where last on-wire byte is the frame type.
+    MalformedFrame       = -24,
+
+    /// QPACK Header block decompression failure.
+    QpackDecompressionFailed = -25,
+
+    /// QPACK encoder stream error.
+    QpackEncoderStreamError = -26,
+
+    /// QPACK decoder stream error.
+    QpackDecoderStreamError = -27,
+}
+
+impl Error {
+    pub fn to_wire(self) -> u16 {
+        match self {
+            Error::Done => 0x0,
+            Error::WrongSettingDirection => 0x1,
+            Error::PushRefused => 0x2,
+            Error::InternalError => 0x3,
+            Error::PushAlreadyInCache => 0x4,
+            Error::RequestCancelled => 0x5,
+            Error::IncompleteRequest => 0x6,
+            Error::ConnectError => 0x07,
+            Error::ExcessiveLoad => 0x08,
+            Error::VersionFallback => 0x09,
+            Error::WrongStream => 0xA,
+            Error::LimitExceeded => 0xB,
+            Error::DuplicatePush => 0xC,
+            Error::UnknownStreamType => 0xD,
+            Error::WrongStreamCount => 0xE,
+            Error::ClosedCriticalStream => 0xF,
+            Error::WrongStreamDirection => 0x10,
+            Error::EarlyResponse => 0x11,
+            Error::MissingSettings => 0x12,
+            Error::UnexpectedFrame => 0x13,
+            Error::RequestRejected => 0x14,
+            Error::GeneralProtocolError => 0xFF,
+            Error::MalformedFrame => 0x10,
+
+            Error::QpackDecompressionFailed => 0x20, // TODO: value is TBD
+            Error::QpackEncoderStreamError => 0x21,  // TODO: value is TBD
+            Error::QpackDecoderStreamError => 0x22,  // TODO: value is TBD
+            Error::BufferTooShort => 0x999,
+        }
+    }
+}
+
+impl std::convert::From<super::Error> for Error {
+    fn from(err: super::Error) -> Self {
+        match err {
+            super::Error::Done => Error::Done,
+            super::Error::BufferTooShort => Error::BufferTooShort,
+            _ => Error::GeneralProtocolError,
+        }
+    }
+}
+
+/// An HTTP/3 configuration.
+pub struct Config {
+    num_placeholders: u64,
+    max_header_list_size: u64,
+    qpack_max_table_capacity: u64,
+    qpack_blocked_streams: u64,
+}
+
+impl Config {
+    pub fn new(
+        num_placeholders: u64, max_header_list_size: u64,
```

Comment:
If the parameters are passed in the `new()` method, then you don't need all the `set_*()` ones, or vice versa If you want to keep the set_* ones then the fields of the struct should probably just be `Option`s.

Reply:
Yeah I realised this when using the FFI in C. These parameters must be set and there are only 4 of them (perhaps going down to 3 soon). Since they are simple types I think removing the set_* functions is the correct way to go. We can add new functions for other things in the future as needed, and not modify the new signature.